### PR TITLE
Trigger external tests on changes to bots/

### PR DESCRIPTION
**Pull request by martinpitt**
Ensure that e. g. image refreshes or testvm.py changes don't break external projects. See individual commits for details.

### CODE REVIEW

**Comment by martinpitt** [comment]
I tested this extensively on my own cockpit fork, where I made [two PRs](https://github.com/martinpitt/cockpit/pulls): [PR#2](https://github.com/martinpitt/cockpit/pull/2) does a change to the bots/ directory, and [PR#3](https://github.com/martinpitt/cockpit/pull/3) changes something else. Now I run `GITHUB_BASE=martinpitt/cockpit bots/tests-scan`, yielding:

```
PRIORITY=0009 bots/make-checkout --verbose --base=master pull/3/head d77dda5f72c92147f12f3f73a4491e50016ee754 && TEST_NAME=pull-3-20180709-103838 TEST_REVISION=d77dda5f72c92147f12f3f73a4491e50016ee754 bots/tests-invoke --rebase=master avocado/fedora pull/3/head
```
This is one of the 19 "standard cockpit" tests for verifying the cockpit change itself. This is the only kind of tests for PR #3, as expected.

```
PRIORITY=0008 bots/make-checkout --verbose --base=master pull/2/head de3d7093dcb77614664c94316dd0a3142a21aff4 && TEST_NAME=pull-2-20180709-103838 TEST_REVISION=de3d7093dcb77614664c94316dd0a3142a21aff4 bots/tests-invoke --rebase=master avocado/fedora pull/2/head
```
Same standard tests for PR#2, but in addition it also triggers the three tests of ostree and two tests of starter kit, against this (`--bots-ref=pull/2/head`) PR:
```
PRIORITY=0009 bots/make-checkout --verbose --base=master --repo=cockpit-project/starter-kit --bots-ref=pull/2/head master '' && GITHUB_BASE=cockpit-project/starter-kit TEST_NAME=pull-2-20180709-103838 TEST_REVISION='' bots/tests-invoke --rebase=master --remote=test cockpit/centos-7 master
PRIORITY=0008 bots/make-checkout --verbose --base=master --repo=cockpit-project/starter-kit --bots-ref=pull/2/head master '' && GITHUB_BASE=cockpit-project/starter-kit TEST_NAME=pull-2-20180709-103838 TEST_REVISION='' bots/tests-invoke --rebase=master --remote=test cockpit/fedora-28 master
PRIORITY=0008 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit-ostree --bots-ref=pull/2/head master '' && GITHUB_BASE=cockpit-project/cockpit-ostree TEST_NAME=pull-2-20180709-103838 TEST_REVISION='' bots/tests-invoke --rebase=master --remote=test cockpit/fedora-atomic master
PRIORITY=0008 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit-ostree --bots-ref=pull/2/head master '' && GITHUB_BASE=cockpit-project/cockpit-ostree TEST_NAME=pull-2-20180709-103838 TEST_REVISION='' bots/tests-invoke --rebase=master --remote=test cockpit/rhel-atomic master
PRIORITY=0009 bots/make-checkout --verbose --base=master --repo=cockpit-project/cockpit-ostree --bots-ref=pull/2/head master '' && GITHUB_BASE=cockpit-project/cockpit-ostree TEST_NAME=pull-2-20180709-103838 TEST_REVISION='' bots/tests-invoke --rebase=master --remote=test cockpit/continuous-atomic master
```

You can see the corresponding "pending" statuses in the PRs (linked above).

I also tested this against the real cockpit project. For that I tagged [PR #9613](https://github.com/cockpit-project/cockpit/pull/9613) with the "test-external" label (as none of Cockpit's current PRs are new, and thus it does not scan any of its patches). This time the human readable output:

```
$ bots/tests-scan -v
pull-9613   cockpit/rhel-atomic                   8  (cockpit-project/cockpit-ostree) [bots@pull/9613/head]
pull-9613   cockpit/fedora-atomic                 8  (cockpit-project/cockpit-ostree) [bots@pull/9613/head]
pull-9613   cockpit/fedora-28                     8  (cockpit-project/starter-kit) [bots@pull/9613/head]
pull-9613   cockpit/continuous-atomic             8  (cockpit-project/cockpit-ostree) [bots@pull/9613/head]
pull-9613   cockpit/centos-7                      8  (cockpit-project/starter-kit) [bots@pull/9613/head]
pull-9068   verify/fedora-28          35d33ac    Testing in progress [cockpit-tasks-mrr3b]
pull-9617   avocado/fedora            7600bbe    Testing in progress [cockpit-tasks-mms5c]
pull-9461   verify/ubuntu-stable      ae976f0    Manual testing required
[... more manual tests ...]
```
and again the expected 5 extra tests were triggered there, but nowhere else.

**Comment by martinpitt** [comment]
This still needs work on the tests-invoke side. While it correctly runs tests, it does not yet properly  update the github status.

**Comment by martinpitt** [comment]
I now fixed the tests-invoke and statuses side as well. I took the tests-scan command from above for PR #9613 corresponding to "cockpit/fedora-28@cockpit-project/starter-kit" and ran it in a local cockpit/tests container with this change, and it [ran successfully](https://fedorapeople.org/groups/cockpit/logs/pull-9613-20180709-125050-33bc09ec-cockpit-project-starter-kit-cockpit-fedora-28/log). 

Note that there are two other bogus "cockpit/fedora-28" and "cockpit/fedora-28@cockpit-project/cockpit" tasks which were from previous buggy test invocations. Please ignore these.

File: bots/tests-scan [hunk 1]
```diff
@@ -51,6 +51,9 @@ REDHAT_VERIFY = {
     'selenium/explorer': [ 'master' ],
 }
 
+# Label: should a PR trigger external tests
+TRIGGER_EXTERNAL = "trigger-external"
```
Comment (stefwalter) [inline bots/tests-scan]: Label should be more descriptive. Perhaps "test-other" or "test-broadly"

File: bots/tests-scan [hunk 2]
```diff
@@ -62,7 +65,10 @@ import urllib.request, urllib.parse, urllib.error
 
 sys.dont_write_bytecode = True
 
-from task import github, REDHAT_PING
+from task import github, label, REDHAT_PING
+
+with open(os.path.join(os.path.dirname(sys.argv[0]), "external.json")) as f:
+    EXTERNAL_PROJECTS = json.load(f)
```
Comment (stefwalter) [inline bots/tests-scan]: Again, I believe this should be included, and not YAC (yet another config).

File: bots/tests-scan [hunk 3]
```diff
@@ -270,17 +296,52 @@ def cockpit_tasks (api, update, policy):
...
+            if repo:  # already a non-cockpit project
+                return False
+            if TRIGGER_EXTERNAL in labels():  # already checked before?
+                return True
+
+            if not have:
+                # this is the first time tests-scan looks at a PR, so determine if it changes bots/
+                with urllib.request.urlopen(pull["patch_url"]) as f:
+                    # enough to look at the git commit header, it lists all changed files
+                    if b"bots/" in f.read(4000):
```
Comment (stefwalter) [inline bots/tests-scan]: This should also be ```src/``` ... Any changes to ```src/``` can be expected to affect external projects.

File: bots/external.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply bots/external.json]: I needed to move it out of .tasks, so this was my initial approach; originally I thought I might need the list somewhere else too, such as image-trigger perhaps; but it turned out that at least for this purpose I don't. So if we are happy with changing code instead the much more rigid JSON, I can put this into tests-scan; I don't mind much either way, so I'll do that.

File: bots/tests-scan (hunk 1 above)
Comment (martinpitt) [reply bots/tests-scan]: Hmm, "test-external"?

Comment (stefwalter) [reply bots/tests-scan]: Sure.

File: bots/tests-scan (hunk 3 above)
Comment (martinpitt) [reply bots/tests-scan]: They don't right now. The only thing external projects check out from cockpit is bots/ (and tests/, but they pin that to a particular tag, so testing PRs won't change things). External projects install the distribution cockpit packages. Perhaps in the future we can build VMs with cockpit from a cockpit PR and somehow inject that into the test suite of starter-kit-like projects, but there are still some steps to go for that. So right now it doesn't make sense to include src/.

Comment (stefwalter) [reply bots/tests-scan]: Makes sense. I guess that's a follow up? To install a custom Cockpit when we change src/ ?

Comment (martinpitt) [reply bots/tests-scan]: Maybe some day, but not immediately after. This requires a lot of infrastructure work, thinking about how to store and pass VMs between bots or how to test multiple projects in one test run, and is generally rather hairy. Also, the benefits of that are relatively marginal - we have to ensure we don't break Cockpit's API, and we must do that already within Cockpit (like we currently do with the rhel-7-5 image). If we break the API and don't notice, and rely on external project tests to tell us, we have already failed IMHO.

Comment (stefwalter) [reply bots/tests-scan]: Okay, but we currently use these exact same tests to tell us when we break API ... and we are moving those tests external. So we are increasing the gap. We currently rely on ourselves to figure out that we broke API, and we're diluting that capability.

Lets be very aware of this, and perhaps increase validation of our API?

**Pull request merged**